add schema.org json-ld metadata to published project pages. closes #147

tompollard opened the pull request:
This change adds schema.org (https://schema.org/) metadata to the published project pages. This is important for indexing in search engines, Google Dataset search, etc.

There are various different metadata formats to choose from (e.g. Microdata, RDFa, or JSON-LD). JSON-LD seems to be most widely recommended (including W3C and Google), so I went for that.

To view the metadata, navigate to a published project (e.g. http://localhost:8000/content/demoeicu/2.0.0/) and then view the source (e.g. view-source:http://localhost:8000/content/demoeicu/2.0.0/).

Useful references/links:

- https://search.google.com/structured-data/testing-tool (tool for testing output)
- https://www.google.com/webmasters/markup-helper/ (tool for manual creation of JSON-LD)
- https://blog.datacite.org/schema-org-register-dois/
- https://developers.google.com/search/docs/guides/intro-structured-data

Sample metadata for: http://localhost:8000/content/demoeicu/2.0.0/

```json
<script type="application/ld+json">
{
  "@context": "https://schema.org/",
  
  "@type": "Dataset",
  
  "name": "Demo eICU Collaborative Research Database",
  "description": "The eICU Collaborative Research Database is a large multi-center critical care database made available by Philips Healthcare in partnership with the MIT Laboratory for Computational Physiology.",
  "version": "2.0.0",
  "license": "PhysioNet Credentialed Health Data License 1.5.0",
  "datePublished" : "Nov. 14, 2018",
  "url": "https://localhost:8000/content/demoeicu/",
  
  "identifier": "https://doi.org/10.13026/G2F309",
  
  "creator": [
  
    {
      "@type": "Person",
      "givenName": "Roger Greenwood",
      "familyName": "Mark",
      "name": "Roger Greenwood Mark"
    }
  
    ],
  "includedInDataCatalog":{
     "@type": "DataCatalog",
     "name": "physionet.org"
  },
  "distribution": [
    {
      "@type": "DataDownload",
      "contentUrl": "https://localhost:8000/content/demoeicu/#files-panel"
    }
  ]
}
</script>
```

bemoody commented:
Thanks.  This sounds useful but I have a lot of questions.

Where is the spec that defines what all of these fields and values
mean?

> <script type="application/ld+json">

Should this perhaps be placed at the *bottom* of the body, rather than
in the head, so it doesn't slow down page loading?

It seems unfortunate that this is required to be embedded in the
document, rather than a <link> or similar.  It adds a bit of bloat
that's not useful for the vast majority of user agents.

> {
>   "@context": "https://schema.org/",
>   {% if project.resource_type == 0 %}{# 0: Data #}
>   "@type": "Dataset",
>   {% elif project.resource_type == 1 %}{# 1: Software #}
>   "@type": "SoftwareSourceCode",
>   {% elif project.resource_type == 2 %}{# 2: Challenge #}
>   "@type": "Article",
>   {% elif project.resource_type == 3 %}{# 3: Model #}
>   "@type": "@type": "Dataset",

This is a typo

>   {% else %}{# >3: treat as data}
>   "@type": "Dataset",
>   {% endif %}

I would suggest that you define a function in
project.models.Metadata that maps from the project resource
type (and/or other, yet-to-be-defined project metadata fields) to
a "@type" value, rather than defining this sort of thing in
template files.

>   "name": "{{ project.title }}",

This value, and similar values below, needs to be escaped:
{{ project.title|escapejs }} should work.  (Okay, I think we don't
actually allow quote characters in titles currently, but we *might*.
Obviously quotes are allowed in the abstract and such.)

>   "description": "{{ project.abstract|truncatechars_html:5000|striptags }}",

Should this be the short description?  Or perhaps the full abstract.
I don't think truncating the abstract is appropriate.

>   "version": "{{ project.version }}",
>   "license": "{{ project.license }}",
>   "datePublished" : "{{ project.publish_datetime|date }}",
>   "url": "https://{{ current_site }}{% url 'published_project_latest' project.slug %}",

Surely this should point to the *current* project URL, not the latest?

>   {% if project.doi %}
>   "identifier": "https://doi.org/{{ project.doi }}",
>   {% endif %}
>   "creator": [
>   {% for author in authors %}
>     {
>       "@type": "Person",

For legacy projects, some authors may not be individuals; I don't know
how you want to address that, but keep it in mind

>       "givenName": "{{ author.first_names }}",
>       "familyName": "{{ author.last_name }}",
>       "name": "{{ author.first_names }} {{ author.last_name }}"
>     }{% if forloop.counter < authors|length %},{% endif %}
>   {% endfor %}
>     ],
>   "includedInDataCatalog":{
>      "@type": "DataCatalog",
>      "name": "physionet.org"

What's a data catalog and what does this mean?
Should this be {{ current_site }}?  (see also issue #595)

>   },
>   "distribution": [
>     {
>       "@type": "DataDownload",
>       "contentUrl": "https://{{ current_site }}{% url 'published_project_latest' project.slug %}#files-panel"

Is this supposed to be machine-readable in some way?  If it's just
meant for human readers then I think I would link to #files rather
than #files-panel.

(Though now I notice that we have two #files-es.  Should fix that.)

Again, you want to link to the current URL, not the latest.

>     }
>   ]
> }
> </script>
>
>
>

Please clean up the extraneous whitespace at the end of the file.

tompollard commented:
Thanks for the helpful comments Benjamin. I've responded to your points below:

> Where is the spec that defines what all of these fields and values mean?

The spec is defined here: https://schema.org/Thing. Start at Thing and work downwards. For example, the "Dataset" type is described here: https://schema.org/Dataset

> Should this perhaps be placed at the *bottom* of the body, rather than in the head, so it doesn't slow down page loading?

Good idea, now included just above `</body>`

> I would suggest that you define a function in project.models.Metadata that maps from the project resource type

I agree, this is a better idea. I've added `schema_org_resource_type()` method to `project.models.Metadata`

> This value, and similar values below, needs to be escaped: {{ project.title|escapejs }}

I added the escapes...perhaps a bit overzealous with them? Thought it was better to be safe!

> Should this be the short description?  Or perhaps the full abstract. I don't think truncating the abstract is appropriate.

The description field must be between 50 and 5000 characters, so I can't think of a better solution than truncating the abstract. In most cases, it won't be truncated, but I think we should have the limit there to be safe. Try testing a shorter description with https://search.google.com/structured-data/testing-tool and you'll get the following error: `Functional . (String length of description must be in range [50, 5000])`, which I think rules out using the `short_description`.

> Surely this should point to the *current* project URL, not the latest?

Good point, now fixed.

> For legacy projects, some authors may not be individuals; I don't know how you want to address that, but keep it in mind

For legacy projects, the creator field will be empty. An empty creator field is not ideal, but it is valid according to the testing tool.

> What's a data catalog and what does this mean?

See: https://developers.google.com/search/docs/data-types/dataset for an explanation. "Datasets are often published in repositories that contain many other datasets. The same dataset can be included in more than one such repository. You can refer to a data catalog that this dataset belongs to by referencing it directly."

This still leaves things a bit fuzzy for me, but my understanding is that the data catalog in our case is PhysioNet (the repository where the dataset lives). Google provides the following "real world" example of metadata for: https://catalog.data.gov/dataset/ncdc-storm-events-database, which suggests to me that "physionet.org" is an acceptable response.

```
{
  "@context":"https://schema.org/",
  "@type":"Dataset",
  "name":"NCDC Storm Events Database",
  "description":"Storm Data is provided by the National Weather Service (NWS) and contain statistics on...",
  "url":"https://catalog.data.gov/dataset/ncdc-storm-events-database",
  "sameAs":"https://gis.ncdc.noaa.gov/geoportal/catalog/search/resource/details.page?id=gov.noaa.ncdc:C00510",
  "identifier": ["https://doi.org/10.1000/182",
                 "https://identifiers.org/ark:/12345/fk1234"],
  "keywords":[
     "ATMOSPHERE > ATMOSPHERIC PHENOMENA > CYCLONES",
     "ATMOSPHERE > ATMOSPHERIC PHENOMENA > DROUGHT",
     "ATMOSPHERE > ATMOSPHERIC PHENOMENA > FOG",
     "ATMOSPHERE > ATMOSPHERIC PHENOMENA > FREEZE"
  ],
  "creator":{
     "@type":"Organization",
     "url": "https://www.ncei.noaa.gov/",
     "name":"OC/NOAA/NESDIS/NCEI > National Centers for Environmental Information, NESDIS, NOAA, U.S. Department of Commerce",
     "contactPoint":{
        "@type":"ContactPoint",
        "contactType": "customer service",
        "telephone":"+1-828-271-4800",
        "email":"ncei.orders@noaa.gov"
     }
  },
  "includedInDataCatalog":{
     "@type":"DataCatalog",
     "name":"data.gov"
  },
  "distribution":[
     {
        "@type":"DataDownload",
        "encodingFormat":"CSV",
        "contentUrl":"http://www.ncdc.noaa.gov/stormevents/ftp.jsp"
     },
     {
        "@type":"DataDownload",
        "encodingFormat":"XML",
        "contentUrl":"http://gis.ncdc.noaa.gov/all-records/catalog/search/resource/details.page?id=gov.noaa.ncdc:C00510"
     }
  ],
  "temporalCoverage":"1950-01-01/2013-12-18",
  "spatialCoverage":{
     "@type":"Place",
     "geo":{
        "@type":"GeoShape",
        "box":"18.0 -65.0 72.0 172.0"
     }
  }
}
```

> Is this supposed to be machine-readable in some way?  If it's just meant for human readers then I think I would link to #files rather than #files-panel.

Ideally both I guess, but I think human readable is the priority, so I've updated the link to `#files`.

bemoody commented:
Thanks for digging into this!

For the demo projects, the Google site reports some errors and warnings:

Article (demobsn):
 - Warning: The property includedInDataCatalog is not recognized by
Google for an object of type Article.
 - Warning: The property distribution is not recognized by Google for
an object of type Article.
 - Error: A value for the author field is required.
 - Error: A value for the headline field is required.
 - Error: A value for the image field is required.
 - Error: A value for the publisher field is required.
 - Warning: The dateModified field is recommended. Please provide a
value if available.
 - Warning: The headline field is recommended. Please provide a value
if available.
 - Warning: The image field is recommended. Please provide a value if available.
 - Warning: The mainEntityOfPage field is recommended. Please provide
a value if available.

SoftwareSourceCode (demoecg)

- Warning: The property includedInDataCatalog is not recognized by
Google for an object of type SoftwareSourceCode.
- Warning: The property distribution is not recognized by Google for
an object of type SoftwareSourceCode.

Dataset (mitdb)

- Error: Open Data Commons Attribution License v1.0 (Text is not a
known valid target type for the license property.)
- Warning: The encodingFormat field is recommended. Please provide a
value if available.

https://schema.org/license indicates that the license should be a URL.

https://schema.org/contentUrl makes me think that is supposed to be a
URL for the *entire* resource (i.e., the zip file or something.)  But
it seems like "distribution" is optional anyway.

tompollard commented:
Doh, thanks for picking this up. Sounds like the fields will need to be tailored to the data type, so I'll need to think about how best to handle this.

tompollard commented:
These changes address the points mentioned above:

- errors raised by the validator for Article type

I have switched Challenge from "Article" to "Dataset", which sidesteps this issue. Not perfect solution, but I think dataset is as appropriate as article for challenge projects, most of which include at least some data.

- Warnings raised for SoftwareSourceCode (`includedInDataCatalog` and `distribution` is not recognized by Google for an object of type SoftwareSourceCode).

The warnings seem fairly benign, so I think it's safe to ignore them. The equivalent property of "includedInDataCatalog" is perhaps "codeRepository", but I'd prefer to deal with later in a larger piece of work to improve the way metadata profiles are handled.

- Error: Open Data Commons Attribution License v1.0 (Text is not a known valid target type for the license property.)

Oddly the tool didn't raise this error for me, but you're right it should be a URL, so I've made this switch.

- Warning: The encodingFormat field is recommended. Please provide a value if available.

I can't see a simple way of getting the format without scanning the files, so I'd prefer to skip this property and deal with later in the work mentioned above.

bemoody commented:
Looks good to me!
